core/services/gateway/jsonrpccodec.go

```diff
@@ -0,0 +1,100 @@
+package gateway
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// Wrapping/unwrapping Message objects into JSON RPC ones folllowing https://www.jsonrpc.org/specification
+type JsonRPCRequest struct {
+	Version string          `json:"jsonrpc"`
+	Id      string          `json:"id"`
+	Method  string          `json:"method"`
+	Params  json.RawMessage `json:"params,omitempty"`
+}
+
+type JsonRPCResponse struct {
+	Version string          `json:"jsonrpc"`
+	Id      string          `json:"id"`
+	Result  json.RawMessage `json:"result,omitempty"`
+	Error   *JsonRPCError   `json:"error,omitempty"`
+}
+
+// JSON-RPC error can only be sent to users. It is not used for messages between Gateways and Nodes.
+type JsonRPCError struct {
+	Code    int             `json:"code"`
+	Message string          `json:"message"`
+	Data    json.RawMessage `json:"data,omitempty"`
+}
+
+func DecodeRequest(msgBytes []byte) (*Message, error) {
+	var request JsonRPCRequest
+	err := json.Unmarshal(msgBytes, &request)
+	if err != nil {
+		return nil, err
+	}
+	msg, err := DecodeMessage(request.Params)
```

Comment:
See the other comment, you won't need this second step if you implement `json.Marshaler`.